crates/wasm-smith/src/core/code_builder.rs

```diff
@@ -3589,10 +4785,21 @@ fn v128_store(
     u: &mut Unstructured,
     module: &Module,
     builder: &mut CodeBuilder,
-) -> Result<Instruction> {
+    instructions: &mut Vec<Instruction>,
+) -> Result<()> {
     builder.pop_operands(&[ValType::V128]);
     let memarg = mem_arg(u, module, builder, &[0, 1, 2, 3, 4])?;
-    Ok(Instruction::V128Store { memarg })
+    if module.config.disallow_traps() {
+        no_traps::store(
+            Instruction::V128Store { memarg },
+            module,
+            builder,
+            instructions,
+        );
+    } else {
+        instructions.push(Instruction::V128Store { memarg });
+    }
```

Comment:
I’m wondering if this pattern would be better served by making a choice between trapping or non-trapping variants and then generating an instruction that is guaranteed to be either way? Are there any instructions that would fall in the “unknown” territory somewhere?

This sort of split model would allow to easily add/maintain other kinds of generation “modes” as well (e.g. deterministic vs non-deterministic), and I wonder if it would result in better fuzz coverage (the fuzzer would be able to flip between the two broad categories of instructions at runtime on a per-instruction basis)

Reply:
@nagisa that's interesting! I'm wondering how useful the "guaranteed to trap" case would actually be. If we were to create a version of this code that does the opposite (ie. makes every division have a denominator of 0, makes every load/store have an out of bounds address, etc), then I would think any generated programs would trap relatively early and not improve coverage all that much except in testing that traps are handled correctly by the runtime. Maybe a wasm-smith flag to generate programs that are guaranteed to _eventually_ trap would be helpful for improving coverage, since that would allow us to explore whether different program state affects the trap-handling behavior of the runtime.

Either way, maybe we discuss this further on https://github.com/bytecodealliance/wasm-tools/issues/266 (or a new issue)? I think work in that direction could be an improvement on this work, but not immediately necessary.

As far as ease of adding/maintaining other modes, I agree that having this turn into something like the below would be unideal:
```rust
if option_1 {
  generate_instruction();
} else if option_2 {
  generate_instruction_but_slightly_differently();
} else {
  generate_instruction_more_differently();
}
```
That being said, I'm inclined to leave that kind of refactoring for when/if we actually add more modes like this rather than trying to think of a more general way of doing this that may never be needed.

Reply:
> That being said, I think I'm inclined to leave that kind of refactoring for when/if we actually add more modes like this

Of course, that was just me thinking out loud, not necessarily suggesting that this must be implemented here and now ^^ Should have made it clear in my original comment.